lib/bandit/websocket/frame.ex

```diff
@@ -172,19 +172,23 @@ defmodule Bandit.WebSocket.Frame do
   defp mask_and_length(length), do: <<0::1, 127::7, length::64>>
 
   # Note that masking is an involution, so we don't need a separate unmask function
-  # Note too that we assume the mask integer is 32-bits and do not check.
   @spec mask(binary(), integer()) :: binary()
-  def mask(payload, mask_integer) do
-    payload_size = byte_size(payload)
-    count_4_bytes = div(payload_size, 4)
+  def mask(payload, mask) when is_binary(payload) and is_integer(mask) and mask >= 0x00000000 and mask <= 0xFFFFFFFF do
```

Comment:
do `is_binary(payload) and is_integer(mask)` add any performance benefits?
at least the integer comparisons on `mask` will fail if it is not an integer, anyway

Reply:
okay the microbenchmarking i'm doing the extra guards range from not making a difference to slightly helping, so, nice.

Reply:
It saves some work by doing the checks once in mask/2, rather than on every iteration of mask/3, similar to the examples in https://www.erlang.org/blog/more-optimizations/.

For example without checks the main `bxor` generates this bytecode:

```erlang
    {line,[{location,"lib/bandit/websocket/frame.ex",181}]}.
    {gc_bif,'bxor',
            {f,0},
            4,
            [{tr,{x,1},{t_integer,{0,4294967295}}},{x,2}],
            {x,1}}.
```

which yields arm64 with type checks and branches:

```asm
# i_bxor_jIssd
    eor x0, x26, x27
    orr x0, x0, 15
    and x8, x26, x27
    and x8, x8, 15
    cmp x8, 15
    b.eq L307
    mov x1, x26
    mov x2, x27
    bl L309
L307:
    mov x26, x0
```

Now with the checks, the compiler knows that register `{x,2}` is always a small integer:

```erlang
    {line,[{location,"lib/bandit/websocket/frame.ex",181}]}.
    {gc_bif,'bxor',
            {f,0},
            4,
            [{tr,{x,3},{t_integer,{0,4294967295}}},
             {tr,{x,2},{t_integer,{0,4294967295}}}],
            {x,3}}.
```

so it can directly xor it:

```asm
# i_bxor_jIssd
# skipped test for small operands because they are always small
    eor x28, x28, x27
    orr x28, x28, 15
```

I think the difference is not huge, maybe because the bytestring append dominates, but I hope that only gets better :)

(I used https://github.com/michalmuskala/decompile to dump bytecode, and `elixir --erl "+JDdump true" lib/bandit/websocket/frame.ex` to get JIT assembly)